frontend/utils/distribution/monthly-event-parser.js

```diff
@@ -0,0 +1,241 @@
+  let startCycleEvent = { data: { monthlyDistribution: [], completedPayments: [] } }
+  let lastStartCycleEvent = {}
+  let monthlyDistribution = [] // For each cycle's monthly distribution calculation
+  let completedMonthlyPayments = [] // For accumulating the payment events of each month's cycle.
+
+  // Create a helper function for calculating each cycle's payment distribution:
+  const paymentsDistribution = function (groupMembers, minCome) {
+    const groupIncomes = groupMembers.map((user) => {
+      return {
+        name: user.name,
+        amount: minCome + user.haveNeed
+      }
+    })
+    return incomeDistribution(groupIncomes, minCome)
+  }
+
+  const handleCycleEvent = (event, artificialEndEvent = false) => {
+    lastStartCycleEvent = JSON.parse(JSON.stringify(startCycleEvent))
+    monthlyDistribution = paymentsDistribution(groupMembers, minCome).map((payment) => {
+      payment.total = payment.amount
+      return payment
+    })
+
+    // Check if it is the last event (the next month after monthstamps cycle event), or if the
+    // final distribution should be adjusted, anyway:
+    monthlyDistribution = subtractDistributions(monthlyDistribution, completedMonthlyPayments)
+
+    monthlyDistribution.forEach((v) => {
+      v.partial = (v.total !== v.amount)
+    })
+
+    startCycleEvent = event
+    startCycleEvent.data.monthlyDistribution = monthlyDistribution
+    startCycleEvent.data.completedPayments = completedMonthlyPayments
+
+    completedMonthlyPayments = [] // and the monthly payments, too...
+  }
+
+  const handleIncomeEvent = (event) => {
+    const oldUser = getUser(event.data.name)
+    if (oldUser) {
+      oldUser.haveNeed = event.data.haveNeed
+    } else {
+      // Add the user who declared their income to our groupMembers list variable
+      groupMembers.push({
+        name: event.data.name,
+        haveNeed: event.data.haveNeed
+      })
+    }
+  }
+
+  const handlePaymentEvent = (event) => {
+    completedMonthlyPayments.push({
+      from: event.data.from,
+      to: event.data.to,
+      amount: event.data.amount,
+      total: 0
+    })
+  }
+
+  const handleExitEvent = (event) => {
+    groupMembers = groupMembers.filter((v) => { return v.name !== event.data.name })
+  }
+
+  // Loop through the events, pro-rating each user's monthly pledges/needs:
+  distributionEvents.forEach((event) => {
+    if (event.type === 'startCycleEvent') {
+      handleCycleEvent(event)
+    } else if (event.type === 'haveNeedEvent') {
+      handleIncomeEvent(event)
+    } else if (event.type === 'paymentEvent') {
+      handlePaymentEvent(event)
+    } else if (event.type === 'userExitsGroupEvent') {
+      handleExitEvent(event)
+    }
+  })
+
+  const lastWhen = distributionEvents[distributionEvents.length - 1].data.when
+
+  const artificialEnd = {
+    type: 'startCycleEvent',
+    data: {
+      when: lastWhen,
+      monthlyDistribution: [] // List to be populated later, by the events-parser
+    }
```

Comment:
Flow complains that `completedPayments` property is missing in `data`

Reply:
Looks like this section of code has been updated, @snowteamer .

Let me know if you spot anything like this in future.